core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,496 @@
+   * cancellation from its parent.
+   */
+  public CancellableContext withCancellation() {
+    return new CancellableContext(this);
+  }
+
+  /**
+   * Create a new context which cannot be cancelled and will not propagate the cancellation of its
+   * parent.
+   */
+  public Context withoutCancellation() {
+    return new Context(this);
+  }
+
+  /**
+   * Create a new context with the given key value set. The new context will cascade cancellation
+   * from its parent.
+   */
+  public <V> Context withValue(Key<V> k1, V v1) {
+    return new Context(this, new Object[][]{{k1, v1}});
+  }
+
+  /**
+   * Create a new context with the given key value set. The new context will cascade cancellation
+   * from its parent.
+   */
+  public <V1, V2> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2) {
+    return new Context(this, new Object[][]{{k1, v1}, {k2, v2}});
+  }
+
+  /**
+   * Create a new context with the given key value set. The new context will cascade cancellation
+   * from its parent.
+   */
+  public <V1, V2, V3> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2, Key<V3> k3, V3 v3) {
+    return new Context(this, new Object[][]{{k1, v1}, {k2, v2}, {k3, v3}});
+  }
+
+  /**
+   * Attach this context to the thread and make it {@link #current}, the previously current context
+   * will be restored when detach is called. It is allowed to attach contexts where
+   * {@link #isCancelled()} is {@code true}.
+   */
+  public void attach() {
+    contextStack.get().addLast(this);
+  }
+
+  /**
+   * Detach the current context from the thread and restore the context that was previously
+   * attached to the thread as the 'current' context.
+   *
+   * @throws java.lang.IllegalStateException if this context is not {@link #current()}.
+   */
+  public void detach() {
+    ArrayDeque<Context> stack = contextStack.get();
+    if (stack.isEmpty() && this == ROOT) {
+      // trying to detach root so just ignore.
+      return;
+    }
+    if (stack.peekLast() != this) {
+      throw new IllegalStateException("Cannot detach a context that is not current");
+    }
+    stack.removeLast();
+  }
+
+  /**
+   * Is this context cancelled.
+   */
+  public boolean isCancelled() {
+    if (parent == null || !cascadesCancellation) {
+      return false;
+    } else {
+      return parent.isCancelled();
+    }
+  }
+
+  /**
+   * Return the cause of the cancellation or {@code null} if context was cancelled without a cause.
+   */
+  @Nullable
+  public Throwable cause() {
+    if (parent == null || !cascadesCancellation) {
+      return null;
+    } else {
+      return parent.cause();
+    }
+  }
+
+
+  /**
+   * Add a listener that will be notified when the context becomes cancelled.
+   */
+  public void addListener(final CancellationListener cancellationListener,
+                          final Executor executor) {
+    Preconditions.checkNotNull(cancellationListener);
+    Preconditions.checkNotNull(executor);
+    if (canBeCancelled()) {
```

Comment:
Then I would favor saving the original `Context` when we add the listener to the `CancellationContext`, which I guess we would have to do in order for `CancellationListener.cancelled()` to pass the correct context.

I will note that the current code doesn't work, because 'Contexts' are guaranteed to be leaked because a context never removes the listener from its parent. That could be fixed, but I'd prefer not to see the creation of CancellationListener for each Context in the parent chain.